bin/cylc-show-times

```diff
@@ -0,0 +1,372 @@
+            [len(h) for h in header]
+        )
+    ]
+    formatter = ' '.join(['%%-%ds' % l for l in max_lengths]) + '\n'
+    sio.write(formatter % header)
+    for r in rows:
+        sio.write(formatter % r)
+    sio.seek(0)
+    return sio
+
+
+def _get_dao(suite):
+    """Return the DAO (public) for suite."""
+    suite_log_dir = GLOBAL_CFG.get_derived_host_item(
+        suite, 'suite log directory'
+    )
+    pub_db_path = os.path.join(
+        os.path.dirname(suite_log_dir),
+        CylcSuiteDAO.DB_FILE_BASE_NAME
+    )
+    return CylcSuiteDAO(pub_db_path, is_public=True)
+
+
+class TimingSummary(object):
+    """Base class for summarizing timing output from cylc run database."""
+
+    def __init__(self, filepath_or_buffer=None):
+        """Set up internal dataframe storage for time durations."""
+
+        self._check_imports()
+        if filepath_or_buffer is not None:
+            self.read_timings(filepath_or_buffer)
+        else:
+            self.df = None
+            self.by_host_and_batch = None
+
+    def read_timings(self, filepath_or_buffer):
+        """
+        Set up time duration dataframe storage based on start/stop/succeed
+        times from flat database output.
+
+        """
+        import pandas as pd
+        df = pd.read_csv(
+            filepath_or_buffer, delim_whitespace=True, index_col=[0, 1, 2, 3],
+            parse_dates=[4, 5, 6]
+        )
+        self.df = pd.DataFrame({
+            'queue_time': (
+                df['start_time'] - df['submit_time']).apply(self._dt_to_s),
+            'run_time':  (
+                df['succeed_time'] - df['start_time']).apply(self._dt_to_s),
+            'total_time': (
+                df['succeed_time'] - df['submit_time']).apply(self._dt_to_s),
+        })
+        self.by_host_and_batch = self.df.groupby(
+            level=['host', 'batch_system']
+        )
+
+    def write_summary(self, buf=None):
+        """Using the stored timings dataframe, output the data summary."""
+
+        if buf is None:
+            buf = sys.stdout
+        self.write_summary_header(buf)
+        for group, df in self.by_host_and_batch:
+            self.write_group_header(buf, group)
+            df_reshape = df.unstack('name').stack(level=0)
+            df_describe = df.groupby(level='name').describe()
+            if df_describe.index.nlevels > 1:
+                df_describe = df_describe.unstack()  # for pandas < 0.20.0
+            df_describe.index.rename(None, inplace=True)
+            for timing_category in self.df.columns:
+                self.write_category(
+                    buf, timing_category, df_reshape, df_describe
+                )
+        self.write_summary_footer(buf)
+
+    def write_summary_header(self, buf):
+        pass
+
+    def write_summary_footer(self, buf):
+        pass
+
+    def write_group_header(self, buf, group):
+        pass
+
+    def write_category(self, buf, category, df_reshape, df_describe):
+        pass
+
```

Comment:
It can, but one of the reasons I implemented it this way is that I don't need to define every single routine in the inheriting classes.  If they're `@abstractmethod`, I'd need to add subroutines with `pass` in the child classes if it's not used (i.e. `TextTimingSummary`  doesn't use a summary header or footer)  If you'd like, though, I can modify to make it more explicit.